test/common/config/udpa_context_params_test.cc

```diff
@@ -0,0 +1,156 @@
+    bar: "a"
+    baz: 42
+  )EOF",
+                            node);
+  const auto context_params = UdpaContextParams::encode(
+      node, {"cluster", "user_agent_version", "locality.region", "locality.sub_zone"}, {}, {}, {});
+  EXPECT_CONTEXT_PARAMS(context_params, Pair("udpa.node.cluster", "some_cluster"),
+                        Pair("udpa.node.locality.sub_zone", "some_sub_zone"),
+                        Pair("udpa.node.locality.region", "some_region"),
+                        Pair("udpa.node.user_agent_version", "1.2.3"));
+}
+
+// Validate user_agent_build_version renderers.
+TEST(UdpaContextParamsTest, NodeUserAgentBuildVersion) {
+  envoy::config::core::v3::Node node;
+  TestUtility::loadFromYaml(R"EOF(
+  user_agent_build_version:
+    version:
+      major_number: 1
+      minor_number: 2
+      patch: 3
+    metadata:
+      foo: true
+      bar: "a"
+      baz: 42
+  )EOF",
+                            node);
+  const auto context_params = UdpaContextParams::encode(
+      node, {"user_agent_build_version.version", "user_agent_build_version.metadata"}, {}, {}, {});
+  EXPECT_CONTEXT_PARAMS(context_params,
+                        Pair("udpa.node.user_agent_build_version.metadata.bar", "\"a\""),
+                        Pair("udpa.node.user_agent_build_version.metadata.baz", "42"),
+                        Pair("udpa.node.user_agent_build_version.metadata.foo", "true"),
+                        Pair("udpa.node.user_agent_build_version.version", "1.2.3"));
+}
+
+// Validate that resource locator context parameters are pass-thru.
+TEST(UdpaContextParamsTest, ResoureContextParams) {
+  udpa::core::v1::ContextParams resource_context_params;
+  TestUtility::loadFromYaml(R"EOF(
+  params:
+    foo: "\"some_string\""
+    bar: "123"
+    baz: "true"
+  )EOF",
+                            resource_context_params);
+  const auto context_params = UdpaContextParams::encode({}, {}, resource_context_params, {}, {});
+  EXPECT_CONTEXT_PARAMS(context_params, Pair("bar", "123"), Pair("baz", "true"),
+                        Pair("foo", "\"some_string\""));
+}
+
+// Validate client feature capabilities context parameter transform.
+TEST(UdpaContextParamsTest, ClientFeatureCapabilities) {
+  const auto context_params =
+      UdpaContextParams::encode({}, {}, {}, {"some.feature", "another.feature"}, {});
+  EXPECT_CONTEXT_PARAMS(context_params, Pair("udpa.client_feature.another.feature", "true"),
+                        Pair("udpa.client_feature.some.feature", "true"));
+}
+
+// Validate per-resource well-known attributes transform.
+TEST(UdpaContextParamsTest, ResourceWktAttribs) {
+  const auto context_params =
+      UdpaContextParams::encode({}, {}, {}, {}, {{"foo", "1"}, {"bar", "2"}});
+  EXPECT_CONTEXT_PARAMS(context_params, Pair("udpa.resource.foo", "1"),
+                        Pair("udpa.resource.bar", "2"));
+}
+
+// Validate that the precedence relationships in the specification hold.
+TEST(UdpaContextParamsTest, Layering) {
```

Comment:
This test doesn't cover what happens if the resource context params specify the same exact param name as one of the "udpa.node.*" entries.  We should validate that the resource context param will be used in that case.

Reply:
It might not be super clear, but it does. We have the node cluster as `some_cluster` in the `Node`, but then override with `udpa.node.cluster` to `another_cluster`.

Reply:
Ah, you're right -- sorry, I was mis-reading it.

This looks good.